tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs

```diff
@@ -302,5 +207,391 @@ private static StringBuilder FormatList(IEnumerable<object> objs, StringBuilder
+                        {
+                            var sb = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
+                            foreach (var x in enumerableDic)
+                            {
+                                sb.Append($"{getKey(x)}, {getValue(x)}, ");
+                            }
+
+                            if (sb.Length > 0)
+                            {
+                                sb.Remove(sb.Length - 2, 2);
+                            }
+
+                            return StringBuilderCache.GetStringAndRelease(sb);
+                        }
+
+                        Log.Warning("EncodeDictionary: object graph too deep, truncating nesting {Items}", GetItemsAsString());
+                        return ddWafObjectMap;
+                    }
+
+                    if (count > WafConstants.MaxContainerSize)
+                    {
+                        Log.Warning<int>("EncodeList: list too long, it will be truncated, MaxMapOrArrayLength {MaxMapOrArrayLength}", WafConstants.MaxContainerSize);
+                    }
+                }
+
+                var childrenCount = !applySafetyLimits || count < WafConstants.MaxContainerSize ? count : WafConstants.MaxContainerSize;
+                var childrenFromPool = ObjectStructSize * childrenCount < MaxBytesForMaxStringLength;
+                var childrenData = childrenFromPool ? pool.Rent() : Marshal.AllocCoTaskMem(ObjectStructSize * childrenCount);
+
+                if (enumerableDic is IDictionary)
+                {
+                    var typeKVP = typeof(KeyValuePair<TKey, TValue>);
+                    if (typeKVP == typeof(KeyValuePair<string, string>))
+                    {
+                        EnumerateItems<string, string>();
+                    }
+                    else if (typeKVP == typeof(KeyValuePair<string, object>))
+                    {
+                        EnumerateItems<string, object>();
+                    }
+                    else if (typeKVP == typeof(KeyValuePair<string, string[]>))
+                    {
+                        EnumerateItems<string, string[]>();
+                    }
+                    else if (typeKVP == typeof(KeyValuePair<string, List<string>>))
+                    {
+                        EnumerateItems<string, List<string>>();
+                    }
+                    else if (typeKVP == typeof(KeyValuePair<string, JToken>))
+                    {
+                        EnumerateItems<string, JToken>();
+                    }
+                    else
+                    {
+                        EnumerateItems<TKey, TValue>();
+                    }
+
+                    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                    void EnumerateItems<TKeySource, TValueSource>()
+                        where TKeySource : notnull
+                    {
+                        var itemData = childrenData;
```

Comment:
Maybe it's better to pass captured vars as params in order to avoid extra heap allocations?
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions#heap-allocations